### DISCUSSION TITLE
inject type improvements

### DISCUSSION THREAD

**Discussion by miraage** [mobxjs/mobx#3741] [Show and tell]
Hello. I was looking for a way to make inject returned typed component properly. I found a workaround that seems to be working for me. I'd like to ask to review this solution and maybe possibly ship it as default in the mobx-react package.

The proposed change is the return type of the inner function
`T extends IReactComponent<infer P> ? IReactComponent<Omit<P, K[number]>> : never `

It would require some changes to work for decorator, but that's the basic idea.

```ts
import * as React from 'react'
import { inject as mobxInject } from 'mobx-react'
import { IReactComponent } from 'mobx-react/dist/types/IReactComponent'

export function inject<K extends string[]>(
  ...stores: K
): <T extends IReactComponent>(
  target: T
) => T extends IReactComponent<infer P> ? IReactComponent<Omit<P, K[number]>> : never {
  return mobxInject(...stores) as any
}

interface Props {
  user: string
}

interface Injections {
  id: number
}

const Pure: React.FC<Props & Injections> = () => {
  return null
}

const Injected = inject('id')(Pure)

const Another = () => {
  return (
    <div>
      <Injected user="me" />
    </div>
  )
}
```

**Comment by kubk** [comment]
Not sure if it should be fixed, since the recommended way to obtain stores is through React Context: https://github.com/mobxjs/mobx/tree/main/packages/mobx-react#provider-and-inject

It doesn't have such issues with TS
